File: 7-linked-list/merge-k-sorted-list.js
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */

function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

var mergeKLists = function(lists) {
    if (!lists || !lists.length) return null
    while (lists.length > 1) {
        let mergedList = new Array()
        for (let i=0; i<lists.length; i+=2) {
            let l1 = lists[i]
            let l2 = i+1 < lists.length ? lists[i+1]:null
            mergedList.push(mergeLists(l1,l2))
        }
        lists = mergedList
    }
    return lists[0]
};

var mergeLists = function(l1,l2) {
    let dummy = new ListNode()
    let curr = dummy
    while (l1 && l2) {
        if (l1.val <= l2.val) {
            curr.next = l1
            l1 = l1.next
        } else {
            curr.next = l2
            l2 = l2.next
        }
        curr = curr.next
    }
    curr.next = l1 ? l1 : l2
    return dummy.next
}

// var mergeKLists = function(lists) {
//     const hashMap = {}
//     for (const list of lists) {
//         let curr = list
//         while (curr) {
//             hashMap[curr.val] = (hashMap[curr.val] || 0) + 1
//             curr = curr.next
//         }
//     }
//     const sortedHashMap = Object.keys(hashMap).sort((a,b)=>a-b)
//     let dummy = new ListNode(-1)
//     let curr = dummy
//     for (const num of sortedHashMap) {
//         for (let i=0; i<hashMap[num]; i++) {
//             curr.next = new ListNode(Number(num))
//             curr = curr.next
//         }
//     }
//     return dummy.next
// };

File: 7-linked-list/merge-k-sorted-list.py
"""
Leet Code 23: Merge k Sorted Lists
Techniques: linked list manipulation, divide and conquer, hash map, merge sort
Time complexity: O(n log k), where n is the total number of nodes across all lists and k is the number of lists
Space complexity: O(k), for storing the lists in the merge process
"""


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def mergeKLists(self, lists):
        if not lists or len(lists) == 0:
            return None
        while len(lists) > 1:
            mergedList = []
            for i in range(0, len(lists), 2):
                l1 = lists[i]
                l2 = lists[i + 1] if i + 1 < len(lists) else None
                mergedList.append(self.mergeList(l1, l2))
            lists = mergedList
        return lists[0]

    def mergeList(self, l1, l2):
        dummy = ListNode()
        curr = dummy
        while l1 and l2:
            if l1.val < l2.val:
                curr.next = l1
                l1 = l1.next
            else:
                curr.next = l2
                l2 = l2.next
            curr = curr.next
        curr.next = l1 if l1 else l2
        return dummy.next

    # def mergeKLists(self, lists):
    #     hash_map = {}
    #     for i in range(len(lists)):
    #         curr = lists[i]
    #         while curr:
    #             hash_map[curr.val] = hash_map.get(curr.val, 0) + 1
    #             curr = curr.next
    #     sorted_hash_map = sorted(hash_map.keys())
    #     dummy = ListNode(-1)
    #     curr = dummy
    #     for i in sorted_hash_map:
    #         for _ in range(hash_map[i]):
    #             curr.next = ListNode(i)
    #             curr = curr.next
    #     return dummy.next
